### PR TITLE
[56] Implement report API @POST

### PR DESCRIPTION
**Pull request by badnameok**
## What did you do

- Implemented report API including checking whether the job with the ID exists or not.

## Demo

- none

### CODE REVIEW

File: apps/api/src/modules/report/report.repository.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  async createReport(payload: ReportPostData) {
+    const { jobID, userID, reason } = payload
+    await this.prisma.report.create({
+      data: {
+        jobId: jobID,
+        reporterId: userID,
+        reason: reason,
+      },
+    })
```
Comment (palmcm) [inline apps/api/src/modules/report/report.repository.ts:18]: This should return created report to get its id from service, or just return reportId

File: apps/api/src/modules/report/report.service.spec.ts [hunk 2]
```diff
@@ -1,19 +1,38 @@
...
 import { ReportService } from './report.service'
 
 describe('ReportService', () => {
   let service: ReportService
+  let repository: ReportRepository
 
   beforeEach(async () => {
     const module: TestingModule = await Test.createTestingModule({
-      providers: [ReportService],
+      providers: [ReportService, ReportRepository, PrismaService],
```
Comment (palmcm) [inline apps/api/src/modules/report/report.service.spec.ts:13]: Add `JobRepository`

File: apps/api/src/modules/report/report.service.spec.ts [hunk 3]
```diff
@@ -1,19 +1,38 @@
...
   })
+
+  describe('postReport', () => {
+    const postReportDTO = {
+      reason: 'Insert reason',
+    }
+
+    it('should post report correctly', async () => {
+      jest.spyOn(repository, 'createReport').mockResolvedValue()
+
```
Comment (palmcm) [inline apps/api/src/modules/report/report.service.spec.ts:31]: Also Mock `jobRepository.getJobById`

File: apps/api/src/modules/report/report.type.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  jobID: number
+  userID: number
+  reason: string
+
+  constructor(jobID: number, userID: number, reason: string) {
+    this.jobID = jobID
+    this.userID = userID
+    this.reason = reason
+  }
+}
```
Comment (palmcm) [inline apps/api/src/modules/report/report.type.ts:13]:
```suggestion
```
remove this file and in repository just receive the data through 3 parameters

**Review by palmcm — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/report/report.controller.ts [hunk 5]
```diff
@@ -1,31 +1,39 @@
...
 import { ReportService } from './report.service'
 
 @ApiTags('report')
 @Controller('report')
 export class ReportController {
   constructor(private readonly reportService: ReportService) {}
 
-  @Post(':id')
+  @Post('job/:id')
+  @UseAuthGuard()
```
Comment (PKhing) [inline apps/api/src/modules/report/report.controller.ts:25]: Only actor should be able to report? (this endpoint should return 403 for casting and admin)

File: apps/api/src/modules/report/report.repository.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+@Injectable()
+export class ReportRepository {
+  constructor(private prisma: PrismaService) {}
+
+  async createReport(payload: ReportPostData) {
+    const { jobID, userID, reason } = payload
+    await this.prisma.report.create({
+      data: {
+        jobId: jobID,
+        reporterId: userID,
```
Comment (PKhing) [inline apps/api/src/modules/report/report.repository.ts:12]: we should use `jobId` not `jobID` for consistency

File: apps/api/src/modules/report/report.service.spec.ts [hunk 7]
```diff
@@ -1,19 +1,38 @@
...
+      await service.postReport(1, postReportDTO, 1)
```
Comment (PKhing) [inline apps/api/src/modules/report/report.service.spec.ts:38]: should we replace `1` with some constant 
ex. MOCK_USER_ID = 1
It look like magic number here and make me confuse when I read it for the first time

**Review by PKhing — CHANGES_REQUESTED** [review]

**Comment by PKhing** [comment]
Don't forget to merge into dev before open PR, so the reviewer can easily test this

**Review by palmcm — APPROVED** [review]
Might do case that user can only report 1 per job, Other feature will left for next sprint or I might fix it

File: apps/api/src/modules/report/report.repository.ts [hunk 8]
```diff
@@ -1,18 +1,15 @@
...
+  async createReport(jobId: number, userId: number, reason: string) {
     await this.prisma.report.create({
       data: {
-        jobId: jobID,
-        reporterId: userID,
+        jobId: jobId,
+        reporterId: userId,
         reason: reason,
       },
     })
```
Comment (PKhing) [inline apps/api/src/modules/report/report.repository.ts:15]: for cleaner code
```ts
  async createReport(jobId: number, reporterId: number, reason: string) {
    await this.prisma.report.create({
      data: {
        jobId,
        reporterId,
        reason,
      },
    })
  }
  ```

**Review by PKhing — APPROVED** [review]

**Pull request merged**